keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+  However, this also interferes with application deployment and has to be supported by the
+  application.
+
+### Cluster Autoscaler
+
+Accepts a `drain-priority-config` option, which is similar to graceful node shutdown in that it
+gives each priority a shutdown grace period. Also has a `max-graceful-termination-sec` option for
+pod termination and a `max-pod-eviction-time` option after which the eviction is forfeited.
+
+Each pod is first analyzed to see if it is drainable. Part of the logic is similar to kubectl and
+its drain filters (see [Cluster Autoscaler rules](https://github.com/kubernetes/autoscaler/blob/554366f979b11aeb82df335a793e4d7a1acfadb4/cluster-autoscaler/simulator/drainability/rules/rules.go#L50-L77)):
+- Mirror pods are skipped.
+- Terminating pods are skipped.
+- Pods and ReplicaSets/ReplicationControllers without owning controllers are blocking by default
+  (the check can be modified with the `skip-nodes-with-custom-controller-pods` option).
+- System pods (in the `kube-system` namespace) without a matching PDB are blocking by default
+  (the check can be modified with the `skip-nodes-with-system-pods` option).
+- Pods with `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation are blocking.
+- Pods with local storage are blocking unless they have a
+  `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation
+  (the check can be modified with the `skip-nodes-with-local-storage` option, e.g., this check
+  is skipped on [AKS](https://learn.microsoft.com/en-us/azure/aks/cluster-autoscaler?tabs=azure-cli#cluster-autoscaler-profile-settings)).
+- Pods with PDBs that do not have `disruptionsAllowed` are blocking.
+
+This can be enhanced with other rules and overrides.
+
+It uses this logic to first check if all pods can be removed from a node. If not, it will report
+those nodes. Then it will group all pods by a priority and evict them gradually from the lowest
+to highest priority. This may include DaemonSet pods.
+
+### Kubelet
+
+Graceful node shutdown is a part of today's solution for node maintenance. Unfortunately it is not
+possible to rely solely on this feature as a go-to solution for graceful node and workload
+termination.
+
+- The graceful node shutdown feature is not application specific and may disrupt workloads
+  prematurely and lead to data loss.
+- The kubelet implements the shutdown via the systemd inhibitor. If dbus or kubelet is restarted
+  during the node shutdown, the shutdown is not registered and pods are terminated ungracefully.
+  Also, new workloads can get scheduled on the node while the node is shutting down. Cluster admins
+  should therefore plan the maintenance in advance and ensure that pods are gracefully removed
+  before attempting to shut down or restart the machine.
+- The kubelet has no way of detecting ongoing maintenance if the node is restarted in the meantime.
+- Graceful termination of static pods during a shutdown is not possible today. It is also not
```

Comment:
I believe Graceful Node Shutdown is not aware of whether a pod was started from a static template or via some other mechanism that scheduled it on the underlying node, so pods that were started as Static Pods will also be terminated in the same way as any other pods by the Graceful Node Shutdown manager.

Reply:
Not sure if I am reading the code correctly. But it seems to me that the shutdown manager obtains the pods as follows:

https://github.com/kubernetes/kubernetes/blob/a31030543c47aac36cf323b885cfb6d8b0a2435f/pkg/kubelet/kubelet.go#L923

which does not seem to include static/mirror pods.